### ISSUE TITLE
5.7.4 performance regression?

### ISSUE THREAD

**Issue by gverdouw** [groue/GRDB.swift#983]
### What did you do?

Updated from 5.7.3 to 5.7.4

### What did you expect to happen?

No noticeable impact on my application

### What happened instead?

My app become noticeably slower in multiple areas. Further debugging seems to show lots of threads sitting in CaseInsensitiveIdentifier.hash(into:). The main place I notice this is when going through a loop doing a fetch and then a create after.

Is there any obvious gotchas with this? Perhaps I am doing something weird with my records?

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 5.7.4 and 5.8.0
**Installation method:** SPM
**Xcode version:** 12.5
**Swift version:** 5
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** 11.4

**Comment by gverdouw** [comment]
One common thing I notice is the records that I notice this with the most are being fetched with a lot of optional associations. For example something like: 

```swift
MainRecord
            .including(optional: MainRecord.association1)
            .including(optional: MainRecord.association2)
            .including(optional: MainRecord.association3)
            .including(optional: MainRecord.association4)
            .including(optional: MainRecord.association5)
            .including(optional: MainRecord.association6)
            .including(optional: MainRecord.association7
                .including(all: AssociatedRecord7.association))
```

There are even more `including`s in my actual query I have only included a handful here... I'm guessing this kind of thing is causing the hash to be too much?

**Comment by groue** [comment]
Hello @gverdouw,

Performance regressions are an important issue. I'd appreciate if you'd provide a detailed setup, please. There are *too many ways* to use GRDB, and I'm not able to guess yours.

**Comment by gverdouw** [comment]
I'll see if I can put together a minimal reproduction project but it might take a while as my app is quite complicated and I am not really sure exactly what might be causing the issue.

While I'm putting that together is there any kind of info in particular that would be useful about my setup?

**Comment by groue** [comment]
For small reproducing examples, I usually go with something like:

``` swift
// A tmp db
let dbQueue = try DatabaseQueue(path: someTempPath)
try dbQueue.write { db in
    // setup minimal schema
}

// Define minimal set of records
struct Player { ... }
struct Team { ... }
struct Game { ... }

// Populate db with minimal data
try dbQueue.write { db in
    // insert minimal data needed for reproducing the issue
}

// This request used to run much faster
let results = try dbQueue.read { db in
    let request = /* some request */
    try request.fetchAll(db)
}
```

**Comment by gverdouw** [comment]
Thanks @groue

In the meantime this is the nasty SQL being generated (I can barely stand to look at it 😆) from my request

```
"SELECT \"responseSubmission\".*, \"apiItemResponse\".*, \"brakeTestItemResponse\".*, \"buildPlateItemResponse\".*, \"buttonItemResponse\".*, \"carfactsItemResponse\".*, \"compliancePlateItemResponse\".*, \"dataItemResponse\".*, \"dateItemResponse\".*, \"dateTimeItemResponse\".*, \"dropDownItemResponse\".*, \"gpsItemResponse\".*, \"groupItemResponse\".*, \"identifierItemResponse\".*, \"inspectionItemResponse\".*, \"mediaItemResponse\".*, \"numberItemResponse\".*, \"odometerItemResponse\".*, \"photoItemResponse\".*, \"photoRollItemResponse\".*, \"registrationNumberItemResponse\".*, \"roadworthyItemResponse\".*, \"remoteInspectionItemResponse\".*, \"scoreItemResponse\".*, \"serialNumberItemResponse\".*, \"signatureItemResponse\".*, \"tallyItemResponse\".*, \"textItemResponse\".*, \"timeItemResponse\".*, \"vehicleItemResponse\".*, \"verifyBankAccountItemResponse\".*, \"verifyBuildPlateItemResponse\".*, \"verifyCompliancePlateItemResponse\".*, \"verifyIdentityItemResponse\".*, \"verifyOdometerItemResponse\".*, \"verifyVehicleItemResponse\".*, \"yesNoItemResponse\".*, \"locationItemResponse\".* FROM \"responseSubmission\" LEFT JOIN \"apiItemResponse\" ON \"apiItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"brakeTestItemResponse\" ON \"brakeTestItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"buildPlateItemResponse\" ON \"buildPlateItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"buttonItemResponse\" ON \"buttonItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"carfactsItemResponse\" ON \"carfactsItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"compliancePlateItemResponse\" ON \"compliancePlateItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"dataItemResponse\" ON \"dataItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"dateItemResponse\" ON \"dateItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"dateTimeItemResponse\" ON \"dateTimeItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"dropDownItemResponse\" ON \"dropDownItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"gpsItemResponse\" ON \"gpsItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"groupItemResponse\" ON \"groupItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"identifierItemResponse\" ON \"identifierItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"inspectionItemResponse\" ON \"inspectionItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"mediaItemResponse\" ON \"mediaItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"numberItemResponse\" ON \"numberItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"odometerItemResponse\" ON \"odometerItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"photoItemResponse\" ON \"photoItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"photoRollItemResponse\" ON \"photoRollItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"registrationNumberItemResponse\" ON \"registrationNumberItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"roadworthyItemResponse\" ON \"roadworthyItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"remoteInspectionItemResponse\" ON \"remoteInspectionItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"scoreItemResponse\" ON \"scoreItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"serialNumberItemResponse\" ON \"serialNumberItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"signatureItemResponse\" ON \"signatureItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"tallyItemResponse\" ON \"tallyItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"textItemResponse\" ON \"textItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"timeItemResponse\" ON \"timeItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"vehicleItemResponse\" ON \"vehicleItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"verifyBankAccountItemResponse\" ON \"verifyBankAccountItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"verifyBuildPlateItemResponse\" ON \"verifyBuildPlateItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"verifyCompliancePlateItemResponse\" ON \"verifyCompliancePlateItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"verifyIdentityItemResponse\" ON \"verifyIdentityItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"verifyOdometerItemResponse\" ON \"verifyOdometerItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"verifyVehicleItemResponse\" ON \"verifyVehicleItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"yesNoItemResponse\" ON \"yesNoItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" LEFT JOIN \"locationItemResponse\" ON \"locationItemResponse\".\"submissionId\" = \"responseSubmission\".\"guid\" WHERE \"responseSubmission\".\"guid\" = ? LIMIT 1"
```

The easy workaround I can do for now (and maybe for good) is make this a view and use that instead of the giant generated request. I think inside the `DatabaseRegion.union(_:)` its just having to recompute the hash for so many tables and columns its just a bit insane for this case 😓

Screenshot from Instruments > Time Profiler on the same bit of code showing the difference

![Screen Shot 2021-06-03 at 7 58 50 pm](https://user-images.githubusercontent.com/90307/120626674-3f447b80-c4a6-11eb-84c8-0d38fd854459.png)

**Comment by groue** [comment]
Ha, you're definitely onto something, @gverdouw! Indeed 5.7.4 has refactored its handling of case-insensitivity for table names.

And the performance regression is... :scream:

OK, I do not need any extra input, thank you very much 👍 This will be fixed shortly! Meanwhile, I hope you could go back to 5.7.3!

**Comment by groue** [comment]
@gverdouw, I know where is the culprit, but I can't reproduce the very high figures of your left screenshot. I'd like to be able to do so, in order to write an actual performance regression test, and make sure this mistake is fixed for good. Would you please give a last hand?

**Comment by gverdouw** [comment]
@groue, ah yes sorry actually this was recorded on a bit of code which spawns a bunch of these calls in multiple dispatch queues (the actual number depends on information from the backend server, but in this case would have been around ~120) hence the huge numbers on the left. I didn't isolate it to one call so a lot of it could have been caused by waiting on locks or something along those lines?

**Comment by gverdouw** [comment]
If you would like me to check any possible fix on my codebase I would be happy to help also

**Comment by groue** [comment]
> If you would like me to check any possible fix on my codebase I would be happy to help also

I will. But I really need some guidance towards reproduction. Or this issue may just not find any solution. I do ask for your time. You can say no, of course. But please just say it so, so that I stop asking for something that will never come.

**Comment by gverdouw** [comment]
Ah sorry I misinterpreted your request. I will restart working on a reproduction project, though I doubt the numbers will be quite as bad as the screenshots as data involved there would be quite painful to generate from scratch.

**Comment by groue** [comment]
> Ah sorry I misinterpreted your request. I will restart working on a reproduction project, though I doubt the numbers will be quite as bad as the screenshots as data involved there would be quite painful to generate from scratch.

As long as you reproduce something that really looks pathologic (works much better with 5.7.3), I'll have the food I'm after :-) Thank you very much for your help!

**Comment by groue** [comment]
Next GRDB version will ship with a naive fix. Maybe you'll tell me, @gverdouw, how it enhances your setup.

**Comment by gverdouw** [comment]
Any difference is very minimal comparing 5.7.3 and develop, massive improvement in my project compared to 5.7.4. Haven't created the reproduction project sorry

**Comment by groue** [comment]
So you think the issue is fixed on the `development` branch?

**Comment by gverdouw** [comment]
Yes, I would say its fixed

**Comment by groue** [comment]
That's great 👍 Thank you very much for your feedback, it is much appreciated :-)

**Comment by groue** [comment]
The fix has shipped in [v5.9.0](https://github.com/groue/GRDB.swift/tree/v5.9.0).
